WarehouseManagement.Infrastructure/Data/WarehouseManagementDbContext.cs

```diff
@@ -20,12 +30,124 @@ public WarehouseManagementDbContext(DbContextOptions<WarehouseManagementDbContex
         public DbSet<VendorMarker> VendorsMarkers { get; set; } = null!;
         public DbSet<VendorZone> VendorsZones { get; set; } = null!;
         public DbSet<ZoneMarker> ZonesMarkers { get; set; } = null!;
+        public DbSet<EntityChange> EntityChanges { get; set; } = null!;
 
         protected override void OnModelCreating(ModelBuilder modelBuilder)
         {
             modelBuilder.ApplyConfigurationsFromAssembly(Assembly.GetExecutingAssembly());
 
             base.OnModelCreating(modelBuilder);
         }
+
+        public async Task<int> SaveChangesWithLogAsync(
+            CancellationToken cancellationToken = default
+        )
+        {
+            var userId =
+                httpContextAccessor.HttpContext.User.FindFirst(ClaimTypes.NameIdentifier)?.Value
+                ?? "guest";
+            var changes = OnBeforeSaveChanges(userId);
+            var result = await base.SaveChangesAsync(cancellationToken);
+            await OnAfterSaveChangesAsync(changes);
+            return result;
+        }
+
+        private List<EntityChange> OnBeforeSaveChanges(string userId)
+        {
+            ChangeTracker.DetectChanges();
+            var changes = new List<EntityChange>();
+
+            var now = DateTime.UtcNow;
+
+            foreach (
+                var entry in ChangeTracker
+                    .Entries()
+                    .Where(e =>
+                        e.State == EntityState.Added
+                        || e.State == EntityState.Modified
+                        || e.State == EntityState.Deleted
+                    )
+            )
+            {
+                var entityName = entry.Entity.GetType().Name;
+                var entityId = GetEntityId(entry);
+
```

Comment:
if `entityId` is empty shouldnt we just do nothing?

Reply:
Now we check if the entityId is IsNullOrEmpty and if it is we skip that record, becouse we won't have Keyless entities in our DB.